packages/liveblocks-react/src/factory.tsx

```diff
@@ -280,45 +319,44 @@ export function createRoomContext<
     // Note: We'll hold on to the initial value given here, and ignore any
     // changes to this argument in subsequent renders
     const frozen = useInitial({
+      instanceId,
       initialPresence,
       initialStorage,
       unstable_batchedUpdates,
-      shouldInitiallyConnect:
-        shouldInitiallyConnect === undefined
-          ? typeof window !== "undefined"
-          : shouldInitiallyConnect,
+      autoConnect,
     });
 
-    const [room, setRoom] = React.useState<
-      Room<TPresence, TStorage, TUserMeta, TRoomEvent>
-    >(() =>
-      client.enter(roomId, {
+    const [{ room }, setRoomLeavePair] = React.useState(() =>
+      stableEnterRoom(frozen.instanceId, roomId, {
         initialPresence: frozen.initialPresence,
         initialStorage: frozen.initialStorage,
-        shouldInitiallyConnect: frozen.shouldInitiallyConnect,
+        autoConnect: false,
         unstable_batchedUpdates: frozen.unstable_batchedUpdates,
       })
     );
 
     React.useEffect(() => {
-      const room = client.enter<TPresence, TStorage, TUserMeta, TRoomEvent>(
-        roomId,
-        {
-          initialPresence: frozen.initialPresence,
-          initialStorage: frozen.initialStorage,
-          shouldInitiallyConnect: frozen.shouldInitiallyConnect,
-          unstable_batchedUpdates: frozen.unstable_batchedUpdates,
-        }
-      );
+      const pair = stableEnterRoom(frozen.instanceId, roomId, {
+        initialPresence: frozen.initialPresence,
+        initialStorage: frozen.initialStorage,
+        autoConnect: false,
+        unstable_batchedUpdates: frozen.unstable_batchedUpdates,
+      });
+
+      setRoomLeavePair(pair);
+      const { room, leave } = pair;
 
-      setRoom(room);
+      if (frozen.autoConnect) {
```

Comment:
oh, I see. but why does this need to happen here? I only worry that sometime in the future autoConnect takes on some additional meaning inside the client itself, but to the react library it means "just connect when mounted" and if those two meanings diverge it could be inconsistent.

Reply:
I've cleaned this up and documented the rationale for this a bit better in dd3be7b9a1403cacd85db17ff062675eb84f7ff5 (as part of the other PR). Hopefully this makes it a bit more obvious. The `autoConnect` feature was added mainly to support the React SSR use case, and to facilitate people that may want to use Liveblocks in "local only" mode (i.e. when you want to nest the same component tree, but not make it multiplayer).